ugc/storage.cpp

```diff
@@ -1,31 +1,367 @@
+    return {};
+  }
+
+  MemReader r(buf.data(), buf.size());
+  NonOwningReaderSource source(r);
+  UGCUpdate update;
+  Deserialize(source, update);
+  return update;
 }
 
 void Storage::SetUGCUpdate(FeatureID const & id, UGCUpdate const & ugc)
 {
-  m_ugc[id] = ugc;
+  ASSERT_THREAD_CHECKER(m_threadChecker, ());
+  auto const feature = GetFeature(id);
+  CHECK_EQUAL(feature->GetFeatureType(), feature::EGeomType::GEOM_POINT, ());
+  auto const & mercator = feature->GetCenter();
+  feature::TypesHolder th(*feature);
+  th.SortBySpec();
+  auto const type = th.GetBestType();
+  for (auto & index : m_UGCIndexes)
+  {
+    if (type == index.m_type && mercator == index.m_mercator && !index.m_deleted)
+    {
+      index.m_deleted = true;
+      m_numberOfDeleted++;
+      break;
+    }
+  }
+  // TODO: Call Defragmentation().
+
+  UGCIndex index;
+  uint64_t offset;
+  if (!GetUGCFileSize(offset))
+    offset = 0;
+
+  index.m_mercator = mercator;
+  index.m_type = type;
+  index.m_mwmName = id.GetMwmName();
+  index.m_dataVersion = id.GetMwmVersion();
+  index.m_featureId = id.m_index;
+  index.m_offset = offset;
+
+  auto const ugcFilePath = GetUGCFilePath();
+  try
+  {
+    FileWriter w(ugcFilePath, FileWriter::Op::OP_APPEND);
+    Serialize(w, ugc);
+    m_UGCIndexes.emplace_back(move(index));
+  }
+  catch (FileWriter::Exception const & exception)
+  {
+    LOG(LERROR, ("Exception while writing file:", ugcFilePath, "reason:", exception.Msg()));
+    return;
+  }
 }
 
 void Storage::Load()
 {
+  ASSERT_THREAD_CHECKER(m_threadChecker, ());
+  string data;
+  auto const indexFilePath = GetIndexFilePath();
+  try
+  {
+    FileReader r(indexFilePath);
+    r.ReadAsString(data);
+  }
+  catch (FileReader::Exception const & exception)
+  {
+    LOG(LWARNING, ("Exception while reading file:", indexFilePath, "reason:", exception.Msg()));
+    return;
+  }
+
+  DeserializeUGCIndex(data, m_UGCIndexes);
+  for (auto const & i : m_UGCIndexes)
+  {
+    if (i.m_deleted)
+      ++m_numberOfDeleted;
+  }
+}
+
+void Storage::SaveIndex() const
+{
+  ASSERT_THREAD_CHECKER(m_threadChecker, ());
+  if (m_UGCIndexes.empty())
+    return;
+
+  auto const jsonData = SerializeUGCIndex(m_UGCIndexes);
+  auto const indexFilePath = GetIndexFilePath();
+  try
+  {
+    FileWriter w(indexFilePath);
+    w.Write(jsonData.c_str(), jsonData.length());
+  }
+  catch (FileWriter::Exception const & exception)
+  {
+    LOG(LERROR, ("Exception while writing file:", indexFilePath, "reason:", exception.Msg()));
+  }
+}
+
+void Storage::Defragmentation()
+{
+  ASSERT_THREAD_CHECKER(m_threadChecker, ());
+  auto const indexesSize = m_UGCIndexes.size();
+  if (m_numberOfDeleted < indexesSize / 2)
+    return;
+
+  auto const ugcFilePath = GetUGCFilePath();
+  auto const tmpUGCFilePath = ugcFilePath + kTmpFileExtension;
+  FileReader r(ugcFilePath);
+
+  vector<uint8_t> buf;
+  uint64_t maxBufSize;
+  CHECK(GetUGCFileSize(maxBufSize), ());
+  buf.resize(maxBufSize);
```

Comment:
nit: in general, you don't need to keep whole buffer in memory - you can read ugcs one by one and dump them to the tmp file immediately.

Reply:
ok